src/components/subscriptions/edit/EditSubscription.js

```diff
@@ -0,0 +1,228 @@
+import { useTranslation } from "react-i18next";
+import { withStyles } from "@material-ui/core/styles";
+import ids from "../ids";
+import { nonEmptyField } from "components/utils/validations";
+import ErrorTypographyWithDialog from "components/error/ErrorTypographyWithDialog";
+import SubscriptionErrorTypographyWithDialog from "../error/SubscriptionErrorTypographyWithDialog";
+import buildID from "components/utils/DebugIDUtil";
+
+function EditSubscriptionDialog(props) {
+    const { open, onClose, parentId, subscription } = props;
+    const [planTypes, setPlanTypes] = useState([]);
+    const [subscriptionSubmission, setSubscriptionSubmission] = useState(null);
+    const [failureReason, setFailureReason] = useState(null);
+    const [mutateSubscriptionError, setMutateSubscriptionError] =
+        useState(null);
+    const { t } = useTranslation("subscriptions");
+
+    const preProcessPlanTypes = React.useCallback(
+        (data) => {
+            if (data?.result?.length > 0) {
+                setPlanTypes(data.result.map((type) => type.name));
+            }
+        },
+        [setPlanTypes]
+    );
+
+    const { mutate: mutateSubscription } = useMutation(
+        () => postSubscription(subscriptionSubmission),
+        {
+            onSuccess: (data) => {
+                const dataResult = data?.result[0];
+                const isFailed = dataResult?.failure_reason;
+                isFailed ? setFailureReason(isFailed) : setFailureReason(null);
+                setMutateSubscriptionError(isFailed ? dataResult : null);
+                if (!isFailed) {
+                    announce({
+                        text: "Success",
+                    });
+                    onClose();
+                }
+            },
+            onError: (err) => {
+                setMutateSubscriptionError(err);
+                console.log(err);
+            },
+        }
+    );
+
+    const handleSubmit = (values) => {
+        setSubscriptionSubmission(formatSubscriptions(values));
+        mutateSubscription();
+    };
+
+    const onCloseForm = () => {
+        onClose();
+        resetState();
+    };
+
+    const resetState = () => {
+        setMutateSubscriptionError(null);
+        setFailureReason(null);
+        setSubscriptionSubmission(null);
+    };
+
+    useQuery({
+        queryKey: [PLAN_TYPES_QUERY_KEY],
+        queryFn: () => getPlanTypes(),
+        enabled: true,
+        staleTime: Infinity,
+        cacheTime: Infinity,
+        onSuccess: preProcessPlanTypes,
+    });
+
+    return (
+        <>
+            <Formik
+                initialValues={mapPropsToValues(subscription)}
+                onSubmit={handleSubmit}
+                enableReinitialize={true}
+            >
+                {({ handleSubmit, ...props }) => {
+                    return (
+                        <Form>
+                            <DEDialog
+                                {...props}
```

Comment:
I am passing props down to be able to use props.resetForm();

Reply:
OK, in that case there may be some confusion on how the `onClose` anonymous function is getting defined here.

It's actually being defined in the same scope as where the `Form` and `DEDialog` tags are used: inside the anonymous child component of this `Formik` tag.

If you are using an IDE like VSCode, you can place your cursor on the `props` inside that `onClose` callback function, and the IDE should highlight the props above in the `{({ handleSubmit, ...props }) => {` line (which is the start of the definition of the anonymous child component of this `Formik` tag); so the IDE is telling us that these are the same variable in the same scope.

You would only need to spread `...props` into the `DEDialog` tag if that component needed to access any of those props inside `src/components/utils/DEDialog.js`.

You can test this by simply removing `{...props}` from the `<DEDialog>` tag, and the `onClose` callback will still work and reset the form.

As another example, you don't have to define that `onClose` callback as an inline, anonymous function. It could be defined like this, which is semantically equivalent and still within the same scope:
```jsx
            <Formik
                initialValues={mapPropsToValues(subscription)}
                onSubmit={handleSubmit}
                enableReinitialize={true}
            >
                {({ handleSubmit, ...props}) => {
                    const onClose = () => {
                        onCloseForm();
                        props.resetForm();
                    };

                    return (
                        <Form>
                            <DEDialog
                                onClose={onClose}
```

Reply:
Thank you for taking the time to explain this to me. I see what you are saying. I did find that removing {...props} from the <DEDialog> tag and <Button> tag resulted in the form no longer being reset on those actions/events. I believe I should move onClose up to prevent the issue and to prevent using the spread in <DEDialog> and <Button>, as suggested.